node/hyperbahn/handler.js

```diff
@@ -337,3 +369,128 @@ function unadvertise(service) {
 };
+
+function convertHosts(hosts) {
+    var res = [];
+    for (var i = 0; i < hosts.length; i++) {
+        var strs = hosts[i].split(':');
+        var obj = {
+            port: parseInt(strs[1])
+        };
+        strs = strs[0].split('.');
+        obj.ip = {
+            ipv4: parseInt(strs[3]) + (parseInt(strs[2]) << 8) +
+                (parseInt(strs[1]) << 16) + (parseInt(strs[0]) << 24)
+        };
+
+        res.push(obj);
+    }
+
+    return res;
+}
+
+HyperbahnHandler.prototype.discover =
+function discover(self, req, head, body, cb) {
+    var serviceName = body && body.query && body.query.serviceName;
+    if (serviceName.length === 0) {
+        cb(null, {
+            ok: false,
+            body: InvalidServiceName({
+                serviceName: serviceName
+            }),
+            typeName: 'invalidServiceName'
+        });
+        return;
+    }
+
+    var exitNodes = self.egressNodes.exitsFor(serviceName);
+    var exitHosts = Object.keys(exitNodes);
+
+    var myHost = self.channel.hostPort;
+    if (exitHosts.indexOf(myHost) === -1) {
+        // Since Hyperbahn is fully connected to service hosts,
+        // randomly picking one suffices
+        var exit = Math.floor(Math.random() * exitHosts.length);
+        self.sendRelayAsThrift({
+            hostPort: exitHosts[exit],
+            body: body,
+            inreq: req,
+            endpoint: 'Hyperbahn::discover'
+        }, function handleForward(err, resp) {
+            if (err) {
+                self.channel.logger.error('Failed to call discover API on exit node', {
+                    error: err,
+                    serviceName: serviceName,
+                    exitHost: exitHosts[exit]
+                });
+                return cb(err, null);
+            }
+
+            // Need to reconstruct the error
+            if (resp.body.name === 'ThriftException') {
+                cb(null, {
+                    ok: false,
+                    body: NoPeersAvailable({
+                        serviceName: serviceName
+                    }),
+                    typeName: 'noPeersAvailable'
+                });
+            } else {
+                cb(null, resp);
+            }
+        });
+    } else {
+        var svcchan = self.channel.topChannel.subChannels[serviceName];
+        var hosts = [];
+        if (svcchan) {
+            hosts = convertHosts(svcchan.peers.keys());
+        }
+        if (hosts.length === 0) {
+            cb(null, {
+                ok: false,
+                body: NoPeersAvailable({
+                    serviceName: serviceName
+                }),
+                typeName: 'noPeersAvailable'
+            });
+        } else {
+            cb(null, {
+                ok: true,
+                body: {
+                    peers: hosts
+                }
+            });
+        }
+    }
+};
+
+HyperbahnHandler.prototype.sendRelayAsThrift =
```

Comment:
We should not need this helper if we make a retryable request directly on the sub channel.